From 10007a8f5ab77172c598bca33d8deb67ac7518d2 Mon Sep 17 00:00:00 2001
From: Mohamed Abbas <mohamed.abbas@intel.com>
Date: Thu, 14 Aug 2014 15:20:24 -0700
Subject: [PATCH 1/1] efibc: EFI Bootloader reboot reason watchdog

Look for panic caused by kernel watchdog.

Any panic caused by kernel watchdog, we need write watchdog
into EFI variable.

Signed-off-by: Mohamed Abbas <mohamed.abbas@intel.com>
---
 drivers/staging/android/efibc.c |   18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/android/efibc.c b/drivers/staging/android/efibc.c
index ac13115ed263..c159ba5a4a14 100644
--- a/drivers/staging/android/efibc.c
+++ b/drivers/staging/android/efibc.c
@@ -33,8 +33,13 @@
 
 #define REBOOT_REASON_CRASH	"kernel_panic"
 #define REBOOT_REASON_NORMAL	"reboot"
-#define REBOOT_REASON_SHUTDOWN	"shutdowm"
+#define REBOOT_REASON_SHUTDOWN	"shutdown"
 #define REBOOT_REASON_HALT	"halt"
+#define REBOOT_REASON_WATCHDOG	"watchdog"
+
+#define WATCHDOG_KERNEL_H	"Watchdog"
+#define WATCHDOG_KERNEL_S	"softlockup"
+#define WATCHDOG_KERNEL_D	"Software Watchdog"
 
 /*
  * Convert char string to efi_char16_t string. Null byte at end is always
@@ -198,8 +203,15 @@ static int efibc_panic_notifier_call(
 		unsigned long what, void *data)
 {
 	int ret;
-
-	ret = efibc_reboot_reason(what, REBOOT_REASON_CRASH);
+	char *str = data;
+
+	if (str &&
+	    (!strncmp(str, WATCHDOG_KERNEL_H, strlen(WATCHDOG_KERNEL_H)) ||
+	    !strncmp(str, WATCHDOG_KERNEL_S, strlen(WATCHDOG_KERNEL_S)) ||
+	    !strncmp(str, WATCHDOG_KERNEL_D, strlen(WATCHDOG_KERNEL_D))))
+		ret = efibc_reboot_reason(what, REBOOT_REASON_WATCHDOG);
+	else
+		ret = efibc_reboot_reason(what, REBOOT_REASON_CRASH);
 
 	return ret;
 }
-- 
1.7.9.5

